// @coreui/react/dist/esm/components/modal/CModalDialog.js
import { __rest } from '../../node_modules/tslib/tslib.es6.js';
import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import classNames from '../../_virtual/index.js';

const CModalDialog = forwardRef((_a, ref) => {
    var { children, alignment, className, fullscreen, scrollable, size } = _a, rest = __rest(_a, ["children", "alignment", "className", "fullscreen", "scrollable", "size"]);
    return (React.createElement("div", Object.assign({ className: classNames('modal-dialog', {
            'modal-dialog-centered': alignment === 'center',
            [typeof fullscreen === 'boolean'
                ? 'modal-fullscreen'
                : `modal-fullscreen-${fullscreen}-down`]: fullscreen,
            'modal-dialog-scrollable': scrollable,
            [`modal-${size}`]: size,
        }, className) }, rest, { ref: ref }), children));
});
CModalDialog.propTypes = {
    alignment: PropTypes.oneOf(['top', 'center']),
    children: PropTypes.node,
    className: PropTypes.string,
    fullscreen: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.oneOf(['sm', 'md', 'lg', 'xl', 'xxl']),
    ]),
    scrollable: PropTypes.bool,
    size: PropTypes.oneOf(['sm', 'lg', 'xl']),
};
CModalDialog.displayName = 'CModalDialog';

export { CModalDialog };
