// @coreui/react/dist/esm/components/image/CImage.js
import { __rest } from '../../node_modules/tslib/tslib.es6.js';
import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import classNames from '../../_virtual/index.js';

const CImage = forwardRef((_a, ref) => {
    var { align, className, fluid, rounded, thumbnail } = _a, rest = __rest(_a, ["align", "className", "fluid", "rounded", "thumbnail"]);
    return (React.createElement("img", Object.assign({ className: classNames({
            [`float-${align}`]: align && (align === 'start' || align === 'end'),
            'd-block mx-auto': align && align === 'center',
            'img-fluid': fluid,
            rounded: rounded,
            'img-thumbnail': thumbnail,
        }, className) || undefined }, rest, { ref: ref })));
});
CImage.propTypes = {
    align: PropTypes.oneOf(['start', 'center', 'end']),
    className: PropTypes.string,
    fluid: PropTypes.bool,
    rounded: PropTypes.bool,
    thumbnail: PropTypes.bool,
};
CImage.displayName = 'CImage';

export { CImage };
