// @coreui/react/dist/cjs/components/list-group/CListGroup.js
'use strict';

var tslib_es6 = require('../../node_modules/tslib/tslib.es6.js');
var React = require('react');
var PropTypes = require('prop-types');
var index = require('../../_virtual/index.js');

const CListGroup = React.forwardRef((_a, ref) => {
    var { children, as: Component = 'ul', className, flush, layout } = _a, rest = tslib_es6.__rest(_a, ["children", "as", "className", "flush", "layout"]);
    return (React.createElement(Component, Object.assign({ className: index.default('list-group', {
            'list-group-flush': flush,
            [`list-group-${layout}`]: layout,
        }, className) }, rest, { ref: ref }), children));
});
CListGroup.propTypes = {
    as: PropTypes.elementType,
    children: PropTypes.node,
    className: PropTypes.string,
    flush: PropTypes.bool,
    layout: PropTypes.oneOf([
        'horizontal',
        'horizontal-sm',
        'horizontal-md',
        'horizontal-lg',
        'horizontal-xl',
        'horizontal-xxl',
    ]),
};
CListGroup.displayName = 'CListGroup';

exports.CListGroup = CListGroup;
