// @coreui/react/dist/cjs/components/toast/CToast.js
'use strict';

var tslib_es6 = require('../../node_modules/tslib/tslib.es6.js');
var React = require('react');
var PropTypes = require('prop-types');
var index = require('../../_virtual/index.js');
var CToastContext = require('./CToastContext.js');
var useForkedRef = require('../../hooks/useForkedRef.js');
require('@popperjs/core');
var props = require('../../props.js');
var Transition = require('../../node_modules/react-transition-group/esm/Transition.js');

const CToast = React.forwardRef((_a, ref) => {
    var { children, animation = true, autohide = true, className, color, delay = 5000, index: index$1, innerKey, visible = false, onClose, onShow } = _a, rest = tslib_es6.__rest(_a, ["children", "animation", "autohide", "className", "color", "delay", "index", "innerKey", "visible", "onClose", "onShow"]);
    const toastRef = React.useRef(null);
    const forkedRef = useForkedRef.useForkedRef(ref, toastRef);
    const [_visible, setVisible] = React.useState(false);
    const timeout = React.useRef(undefined);
    React.useEffect(() => {
        setVisible(visible);
    }, [visible]);
    const contextValues = {
        visible: _visible,
        setVisible,
    };
    // triggered on mount and destroy
    React.useEffect(() => () => clearTimeout(timeout.current), []);
    React.useEffect(() => {
        _autohide();
    }, [_visible]);
    const _autohide = () => {
        if (autohide) {
            clearTimeout(timeout.current);
            timeout.current = window.setTimeout(() => {
                setVisible(false);
            }, delay);
        }
    };
    return (React.createElement(Transition.default, { in: _visible, nodeRef: toastRef, onEnter: () => onShow && onShow(index$1 !== null && index$1 !== void 0 ? index$1 : null), onExited: () => onClose && onClose(index$1 !== null && index$1 !== void 0 ? index$1 : null), timeout: 250, unmountOnExit: true }, (state) => (React.createElement(CToastContext.CToastContext.Provider, { value: contextValues },
        React.createElement("div", Object.assign({ className: index.default('toast', {
                fade: animation,
                [`bg-${color}`]: color,
                'border-0': color,
                'show showing': state === 'entering' || state === 'exiting',
                show: state === 'entered',
            }, className), "aria-live": "assertive", "aria-atomic": "true", role: "alert", onMouseEnter: () => clearTimeout(timeout.current), onMouseLeave: () => _autohide() }, rest, { key: innerKey, ref: forkedRef }), children)))));
});
CToast.propTypes = {
    animation: PropTypes.bool,
    autohide: PropTypes.bool,
    children: PropTypes.node,
    className: PropTypes.string,
    color: props.colorPropType,
    delay: PropTypes.number,
    index: PropTypes.number,
    innerKey: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    onClose: PropTypes.func,
    onShow: PropTypes.func,
    visible: PropTypes.bool,
};
CToast.displayName = 'CToast';

exports.CToast = CToast;
