// @coreui/react/dist/cjs/components/accordion/CAccordion.js
'use strict';

var tslib_es6 = require('../../node_modules/tslib/tslib.es6.js');
var React = require('react');
var PropTypes = require('prop-types');
var index = require('../../_virtual/index.js');
var CAccordionContext = require('./CAccordionContext.js');

const CAccordion = React.forwardRef((_a, ref) => {
    var { children, activeItemKey, alwaysOpen = false, className, flush } = _a, rest = tslib_es6.__rest(_a, ["children", "activeItemKey", "alwaysOpen", "className", "flush"]);
    const [_activeItemKey, setActiveKey] = React.useState(activeItemKey);
    return (React.createElement("div", Object.assign({ className: index.default('accordion', { 'accordion-flush': flush }, className) }, rest, { ref: ref }),
        React.createElement(CAccordionContext.CAccordionContext.Provider, { value: { _activeItemKey, alwaysOpen, setActiveKey } }, children)));
});
CAccordion.propTypes = {
    activeItemKey: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    alwaysOpen: PropTypes.bool,
    children: PropTypes.node,
    className: PropTypes.string,
    flush: PropTypes.bool,
};
CAccordion.displayName = 'CAccordion';

exports.CAccordion = CAccordion;
