// @coreui/react/dist/cjs/components/navbar/CNavbar.js
'use strict';

var tslib_es6 = require('../../node_modules/tslib/tslib.es6.js');
var React = require('react');
var PropTypes = require('prop-types');
var index = require('../../_virtual/index.js');
var props = require('../../props.js');

const CNavbar = React.forwardRef((_a, ref) => {
    var { children, as: Component = 'nav', className, color, colorScheme, container, expand, placement } = _a, rest = tslib_es6.__rest(_a, ["children", "as", "className", "color", "colorScheme", "container", "expand", "placement"]);
    return (React.createElement(Component, Object.assign({ className: index.default('navbar', {
            [`bg-${color}`]: color,
            [typeof expand === 'boolean' ? 'navbar-expand' : `navbar-expand-${expand}`]: expand,
        }, placement, className) }, (colorScheme && { 'data-coreui-theme': colorScheme }), rest, { ref: ref }), container ? (React.createElement("div", { className: typeof container === 'string' ? `container-${container}` : 'container' }, children)) : (React.createElement(React.Fragment, null, children))));
});
CNavbar.propTypes = {
    as: PropTypes.elementType,
    children: PropTypes.node,
    className: PropTypes.string,
    color: props.colorPropType,
    colorScheme: PropTypes.oneOf(['dark', 'light']),
    container: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.oneOf([
            'sm',
            'md',
            'lg',
            'xl',
            'xxl',
            'fluid',
        ]),
    ]),
    expand: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.oneOf(['sm', 'md', 'lg', 'xl', 'xxl']),
    ]),
    placement: PropTypes.oneOf(['fixed-top', 'fixed-bottom', 'sticky-top']),
};
CNavbar.displayName = 'CNavbar';

exports.CNavbar = CNavbar;
