// @coreui/react/dist/cjs/components/image/CImage.js
'use strict';

var tslib_es6 = require('../../node_modules/tslib/tslib.es6.js');
var React = require('react');
var PropTypes = require('prop-types');
var index = require('../../_virtual/index.js');

const CImage = React.forwardRef((_a, ref) => {
    var { align, className, fluid, rounded, thumbnail } = _a, rest = tslib_es6.__rest(_a, ["align", "className", "fluid", "rounded", "thumbnail"]);
    return (React.createElement("img", Object.assign({ className: index.default({
            [`float-${align}`]: align && (align === 'start' || align === 'end'),
            'd-block mx-auto': align && align === 'center',
            'img-fluid': fluid,
            rounded: rounded,
            'img-thumbnail': thumbnail,
        }, className) || undefined }, rest, { ref: ref })));
});
CImage.propTypes = {
    align: PropTypes.oneOf(['start', 'center', 'end']),
    className: PropTypes.string,
    fluid: PropTypes.bool,
    rounded: PropTypes.bool,
    thumbnail: PropTypes.bool,
};
CImage.displayName = 'CImage';

exports.CImage = CImage;
