// @coreui/react/dist/cjs/components/alert/CAlert.js
'use strict';

var tslib_es6 = require('../../node_modules/tslib/tslib.es6.js');
var React = require('react');
var PropTypes = require('prop-types');
var index = require('../../_virtual/index.js');
var CCloseButton = require('../close-button/CCloseButton.js');
var useForkedRef = require('../../hooks/useForkedRef.js');
require('@popperjs/core');
var props = require('../../props.js');
var Transition = require('../../node_modules/react-transition-group/esm/Transition.js');

const CAlert = React.forwardRef((_a, ref) => {
    var { children, className, color = 'primary', dismissible, variant, visible = true, onClose } = _a, rest = tslib_es6.__rest(_a, ["children", "className", "color", "dismissible", "variant", "visible", "onClose"]);
    const alertRef = React.useRef(null);
    const forkedRef = useForkedRef.useForkedRef(ref, alertRef);
    const [_visible, setVisible] = React.useState(visible);
    React.useEffect(() => {
        setVisible(visible);
    }, [visible]);
    return (React.createElement(Transition.default, { in: _visible, mountOnEnter: true, nodeRef: alertRef, onExit: onClose, timeout: 150, unmountOnExit: true }, (state) => (React.createElement("div", Object.assign({ className: index.default('alert', variant === 'solid' ? `bg-${color} text-white` : `alert-${color}`, {
            'alert-dismissible fade': dismissible,
            show: state === 'entered',
        }, className), role: "alert" }, rest, { ref: forkedRef }),
        children,
        dismissible && React.createElement(CCloseButton.CCloseButton, { onClick: () => setVisible(false) })))));
});
CAlert.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    color: props.colorPropType.isRequired,
    dismissible: PropTypes.bool,
    onClose: PropTypes.func,
    variant: PropTypes.string,
    visible: PropTypes.bool,
};
CAlert.displayName = 'CAlert';

exports.CAlert = CAlert;
