// @coreui/react/dist/cjs/components/table/CTableDataCell.js
'use strict';

var tslib_es6 = require('../../node_modules/tslib/tslib.es6.js');
var React = require('react');
var PropTypes = require('prop-types');
var index = require('../../_virtual/index.js');
var props = require('../../props.js');

const CTableDataCell = React.forwardRef((_a, ref) => {
    var { children, active, align, className, color } = _a, rest = tslib_es6.__rest(_a, ["children", "active", "align", "className", "color"]);
    const Component = rest.scope ? 'th' : 'td';
    return (React.createElement(Component, Object.assign({ className: index.default({
            [`align-${align}`]: align,
            'table-active': active,
            [`table-${color}`]: color,
        }, className) || undefined }, rest, { ref: ref }), children));
});
CTableDataCell.propTypes = {
    active: PropTypes.bool,
    align: PropTypes.oneOf(['bottom', 'middle', 'top']),
    children: PropTypes.node,
    className: PropTypes.string,
    color: props.colorPropType,
};
CTableDataCell.displayName = 'CTableDataCell';

exports.CTableDataCell = CTableDataCell;
