// @coreui/react/dist/esm/components/spinner/CSpinner.js
import { __rest } from '../../node_modules/tslib/tslib.es6.js';
import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import classNames from '../../_virtual/index.js';
import { colorPropType } from '../../props.js';

const CSpinner = forwardRef((_a, ref) => {
    var { as: Component = 'div', className, color, size, variant = 'border', visuallyHiddenLabel = 'Loading...' } = _a, rest = __rest(_a, ["as", "className", "color", "size", "variant", "visuallyHiddenLabel"]);
    return (React.createElement(Component, Object.assign({ className: classNames(`spinner-${variant}`, {
            [`spinner-${variant}-${size}`]: size,
            [`text-${color}`]: color,
        }, className), role: "status" }, rest, { ref: ref }),
        React.createElement("span", { className: "visually-hidden" }, visuallyHiddenLabel)));
});
CSpinner.propTypes = {
    as: PropTypes.string,
    className: PropTypes.string,
    color: colorPropType,
    size: PropTypes.oneOf(['sm']),
    variant: PropTypes.oneOf(['border', 'grow']),
    visuallyHiddenLabel: PropTypes.string,
};
CSpinner.displayName = 'CSpinner';

export { CSpinner };
