// @coreui/react/dist/cjs/components/form/CFormFeedback.js
'use strict';

var tslib_es6 = require('../../node_modules/tslib/tslib.es6.js');
var React = require('react');
var PropTypes = require('prop-types');
var index = require('../../_virtual/index.js');

const CFormFeedback = React.forwardRef((_a, ref) => {
    var { children, as: Component = 'div', className, invalid, tooltip, valid } = _a, rest = tslib_es6.__rest(_a, ["children", "as", "className", "invalid", "tooltip", "valid"]);
    return (React.createElement(Component, Object.assign({ className: index.default({
            [`invalid-${tooltip ? 'tooltip' : 'feedback'}`]: invalid,
            [`valid-${tooltip ? 'tooltip' : 'feedback'}`]: valid,
        }, className) }, rest, { ref: ref }), children));
});
CFormFeedback.propTypes = {
    as: PropTypes.elementType,
    children: PropTypes.node,
    className: PropTypes.string,
    invalid: PropTypes.bool,
    tooltip: PropTypes.bool,
    valid: PropTypes.bool,
};
CFormFeedback.displayName = 'CFormFeedback';

exports.CFormFeedback = CFormFeedback;
