// @coreui/react/dist/esm/components/tabs/CTabPane.js
import { __rest } from '../../node_modules/tslib/tslib.es6.js';
import React, { forwardRef, useRef } from 'react';
import PropTypes from 'prop-types';
import classNames from '../../_virtual/index.js';
import { useForkedRef } from '../../hooks/useForkedRef.js';
import '@popperjs/core';
import Transition from '../../node_modules/react-transition-group/esm/Transition.js';

const CTabPane = forwardRef((_a, ref) => {
    var { children, className, onHide, onShow, transition = true, visible } = _a, rest = __rest(_a, ["children", "className", "onHide", "onShow", "transition", "visible"]);
    const tabPaneRef = useRef(null);
    const forkedRef = useForkedRef(ref, tabPaneRef);
    return (React.createElement(Transition, { in: visible, nodeRef: tabPaneRef, onEnter: onShow, onExit: onHide, timeout: 150 }, (state) => (React.createElement("div", Object.assign({ className: classNames('tab-pane', {
            active: visible,
            fade: transition,
            show: state === 'entered',
        }, className) }, rest, { ref: forkedRef }), children))));
});
CTabPane.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    onHide: PropTypes.func,
    onShow: PropTypes.func,
    transition: PropTypes.bool,
    visible: PropTypes.bool,
};
CTabPane.displayName = 'CTabPane';

export { CTabPane };
