// @coreui/react/dist/esm/components/navbar/CNavbar.js
import { __rest } from '../../node_modules/tslib/tslib.es6.js';
import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import classNames from '../../_virtual/index.js';
import { colorPropType } from '../../props.js';

const CNavbar = forwardRef((_a, ref) => {
    var { children, as: Component = 'nav', className, color, colorScheme, container, expand, placement } = _a, rest = __rest(_a, ["children", "as", "className", "color", "colorScheme", "container", "expand", "placement"]);
    return (React.createElement(Component, Object.assign({ className: classNames('navbar', {
            [`bg-${color}`]: color,
            [typeof expand === 'boolean' ? 'navbar-expand' : `navbar-expand-${expand}`]: expand,
        }, placement, className) }, (colorScheme && { 'data-coreui-theme': colorScheme }), rest, { ref: ref }), container ? (React.createElement("div", { className: typeof container === 'string' ? `container-${container}` : 'container' }, children)) : (React.createElement(React.Fragment, null, children))));
});
CNavbar.propTypes = {
    as: PropTypes.elementType,
    children: PropTypes.node,
    className: PropTypes.string,
    color: colorPropType,
    colorScheme: PropTypes.oneOf(['dark', 'light']),
    container: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.oneOf([
            'sm',
            'md',
            'lg',
            'xl',
            'xxl',
            'fluid',
        ]),
    ]),
    expand: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.oneOf(['sm', 'md', 'lg', 'xl', 'xxl']),
    ]),
    placement: PropTypes.oneOf(['fixed-top', 'fixed-bottom', 'sticky-top']),
};
CNavbar.displayName = 'CNavbar';

export { CNavbar };
