// @coreui/react/dist/cjs/components/nav/CNav.js
'use strict';

var tslib_es6 = require('../../node_modules/tslib/tslib.es6.js');
var React = require('react');
var PropTypes = require('prop-types');
var index = require('../../_virtual/index.js');

const CNav = React.forwardRef((_a, ref) => {
    var { children, as: Component = 'ul', className, layout, variant } = _a, rest = tslib_es6.__rest(_a, ["children", "as", "className", "layout", "variant"]);
    return (React.createElement(Component, Object.assign({ className: index.default('nav', variant === 'enclosed-pills' && 'nav-enclosed', // Enclosed pills variant required for `.nav-enclosed` class
        {
            [`nav-${layout}`]: layout,
            [`nav-${variant}`]: variant,
        }, className), role: "navigation" }, rest, { ref: ref }), children));
});
CNav.propTypes = {
    as: PropTypes.elementType,
    children: PropTypes.node,
    className: PropTypes.string,
    layout: PropTypes.oneOf(['fill', 'justified']),
    variant: PropTypes.oneOf([
        'enclosed',
        'enclosed-pills',
        'pills',
        'tabs',
        'underline',
        'underline-border',
    ]),
};
CNav.displayName = 'CNav';

exports.CNav = CNav;
