// @coreui/react/dist/cjs/components/widgets/CWidgetStatsA.js
'use strict';

var tslib_es6 = require('../../node_modules/tslib/tslib.es6.js');
var React = require('react');
var PropTypes = require('prop-types');
var index = require('../../_virtual/index.js');
var CCard = require('../card/CCard.js');
var CCardBody = require('../card/CCardBody.js');
require('../card/CCardFooter.js');
require('../card/CCardGroup.js');
require('../card/CCardHeader.js');
require('../card/CCardImage.js');
require('../card/CCardImageOverlay.js');
require('../card/CCardLink.js');
require('../card/CCardSubtitle.js');
require('../card/CCardText.js');
require('../card/CCardTitle.js');
var props = require('../../props.js');

const CWidgetStatsA = React.forwardRef((_a, ref) => {
    var { action, chart, className, color, title, value } = _a, rest = tslib_es6.__rest(_a, ["action", "chart", "className", "color", "title", "value"]);
    return (React.createElement(CCard.CCard, Object.assign({ className: index.default({ [`bg-${color}`]: color, 'text-white': color }, className) }, rest, { ref: ref }),
        React.createElement(CCardBody.CCardBody, { className: "pb-0 d-flex justify-content-between align-items-start" },
            React.createElement("div", null,
                value && React.createElement("div", { className: "fs-4 fw-semibold" }, value),
                title && React.createElement("div", null, title)),
            action),
        chart));
});
CWidgetStatsA.propTypes = {
    action: PropTypes.node,
    chart: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    className: PropTypes.string,
    color: props.colorPropType,
    title: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.node, PropTypes.number]),
};
CWidgetStatsA.displayName = 'CWidgetStatsA';

exports.CWidgetStatsA = CWidgetStatsA;
