// @coreui/react/dist/cjs/components/widgets/CWidgetStatsF.js
'use strict';

var tslib_es6 = require('../../node_modules/tslib/tslib.es6.js');
var React = require('react');
var PropTypes = require('prop-types');
var CCard = require('../card/CCard.js');
var CCardBody = require('../card/CCardBody.js');
var CCardFooter = require('../card/CCardFooter.js');
require('../card/CCardGroup.js');
require('../card/CCardHeader.js');
require('../card/CCardImage.js');
require('../card/CCardImageOverlay.js');
require('../card/CCardLink.js');
require('../card/CCardSubtitle.js');
require('../card/CCardText.js');
require('../card/CCardTitle.js');
var props = require('../../props.js');

const CWidgetStatsF = React.forwardRef((_a, ref) => {
    var { className, color, footer, icon, padding = true, title, value } = _a, rest = tslib_es6.__rest(_a, ["className", "color", "footer", "icon", "padding", "title", "value"]);
    return (React.createElement(CCard.CCard, Object.assign({ className: className }, rest, { ref: ref }),
        React.createElement(CCardBody.CCardBody, { className: `d-flex align-items-center ${padding === false && 'p-0'}` },
            React.createElement("div", { className: `me-3 text-white bg-${color} ${padding ? 'p-3' : 'p-4'}` }, icon),
            React.createElement("div", null,
                React.createElement("div", { className: `fs-6 fw-semibold text-${color}` }, value),
                React.createElement("div", { className: "text-body-secondary text-uppercase fw-semibold small" }, title))),
        footer && React.createElement(CCardFooter.CCardFooter, null, footer)));
});
CWidgetStatsF.propTypes = {
    className: PropTypes.string,
    color: props.colorPropType,
    footer: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    icon: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    padding: PropTypes.bool,
    title: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.node, PropTypes.number]),
};
CWidgetStatsF.displayName = 'CWidgetStatsF';

exports.CWidgetStatsF = CWidgetStatsF;
